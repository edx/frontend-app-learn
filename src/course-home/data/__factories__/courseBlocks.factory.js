import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies
import './block.factory';

// Generates an Array of block IDs, either from a single block or an array of blocks.
const getIds = (attr) => {
  const blocks = Array.isArray(attr) ? attr : [attr];
  return blocks.map(block => block.id);
};

// Generates an Object in { [block.id]: block } format, either from a single block or an array of blocks.
const getBlocks = (attr) => {
  const blocks = Array.isArray(attr) ? attr : [attr];
  // eslint-disable-next-line no-return-assign,no-sequences
  return blocks.reduce((acc, block) => (acc[block.id] = block, acc), {});
};

Factory.define('courseBlocks')
  .option('courseId', 'course-v1:edX+DemoX+Demo_Course')
  .option('units', ['courseId'], courseId => ([
    Factory.build(
      'block',
      { type: 'vertical' },
      { courseId },
    ),
  ]))
  .option('sequence', ['courseId', 'units'], (courseId, child) => Factory.build(
    'block',
    { type: 'sequential', children: getIds(child) },
    { courseId },
  ))
  .option('section', ['courseId', 'sequence'], (courseId, child) => Factory.build(
    'block',
    { type: 'chapter', children: getIds(child) },
    { courseId },
  ))
  .option('course', ['courseId', 'section'], (courseId, child) => Factory.build(
    'block',
    { type: 'course', children: getIds(child) },
    { courseId },
  ))
  .attr(
    'blocks',
    ['course', 'section', 'sequence', 'units'],
    (course, section, sequence, units) => ({
      [course.id]: course,
      ...getBlocks(section),
      ...getBlocks(sequence),
      ...getBlocks(units),
    }),
  )
  .attr('root', ['course'], course => course.id);

/**
 * Builds a course with a single chapter, sequence, and unit.
 */
export default function buildSimpleCourseBlocks(courseId, title, options = {}) {
  const sequenceBlock = options.sequenceBlock || [Factory.build(
    'block',
    { type: 'sequential' },
    { courseId },
  )];
  const sectionBlock = options.sectionBlock || Factory.build(
    'block',
    {
      type: 'chapter',
      display_name: 'Title of Section',
      complete: options.complete || false,
      effort_time: 15,
      effort_activities: 2,
      resume_block: options.resumeBlock || false,
      children: sequenceBlock.map(block => block.id),
    },
    { courseId },
  );
  const courseBlock = options.courseBlock || Factory.build(
    'block',
    { type: 'course', display_name: title, children: [sectionBlock.id] },
    { courseId },
  );
  return {
    courseBlocks: options.courseBlocks || Factory.build(
      'courseBlocks',
      { courseId },
      {
        sequence: sequenceBlock,
        section: sectionBlock,
        course: courseBlock,
      },
    ),
    sequenceBlock,
    sectionBlock,
    courseBlock,
  };
}
