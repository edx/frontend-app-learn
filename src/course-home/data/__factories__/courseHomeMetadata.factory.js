import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies

Factory.define('courseHomeMetadata')
  .sequence(
    'courseId', (courseId) => `course-v1:edX+DemoX+Demo_Course_${courseId}`,
  )
  .option('host', 'http://localhost:18000')
  .attrs({
    is_staff: false,
    original_user_is_staff: false,
    number: 'DemoX',
    org: 'edX',
    title: 'Demonstration Course',
    is_self_paced: false,
    is_enrolled: false,
    can_load_courseware: false,
  })
  .attr(
    'tabs', ['courseId', 'host'], (courseId, host) => {
      const tabs = [
        Factory.build(
          'tab',
          {
            title: 'Course',
            priority: 0,
            slug: 'courseware',
            type: 'courseware',
          },
          { courseId, path: 'course/' },
        ),
        Factory.build(
          'tab',
          {
            title: 'Discussion',
            priority: 1,
            slug: 'discussion',
            type: 'discussion',
          },
          { courseId, path: 'discussion/forum/' },
        ),
        Factory.build(
          'tab',
          {
            title: 'Wiki',
            priority: 2,
            slug: 'wiki',
            type: 'wiki',
          },
          { courseId, path: 'course_wiki' },
        ),
        Factory.build(
          'tab',
          {
            title: 'Progress',
            priority: 3,
            slug: 'progress',
            type: 'progress',
          },
          { courseId, path: 'progress' },
        ),
        Factory.build(
          'tab',
          {
            title: 'Instructor',
            priority: 4,
            slug: 'instructor',
            type: 'instructor',
          },
          { courseId, path: 'instructor' },
        ),
        Factory.build(
          'tab',
          {
            title: 'Dates',
            priority: 5,
            slug: 'dates',
            type: 'dates',
          },
          { courseId, path: 'dates' },
        ),
      ];

      return tabs.map(
        tab => ({
          tab_id: tab.slug,
          title: tab.title,
          url: `${host}${tab.url}`,
        }),
      );
    },
  );
