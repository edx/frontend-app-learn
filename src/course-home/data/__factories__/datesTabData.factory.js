import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies

Factory.define('datesTabData')
  .attrs({
    dates_banner_info: {
      content_type_gating_enabled: false,
      missed_gated_content: false,
      missed_deadlines: false,
    },
    course_date_blocks: [
      {
        date: '2013-02-05T05:00:00Z',
        date_type: 'course-start-date',
        description: '',
        learner_has_access: true,
        link: '',
        title: 'Course Starts',
        extraInfo: '',
      },
    ],
    missed_deadlines: false,
    missed_gated_content: false,
    learner_is_full_access: true,
    user_timezone: null,
    verified_upgrade_link: 'http://localhost:18130/basket/add/?sku=8CF08E5',
  });
