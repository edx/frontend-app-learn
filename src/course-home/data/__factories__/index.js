import './courseHomeMetadata.factory';
import './datesTabData.factory';
import './outlineTabData.factory';
