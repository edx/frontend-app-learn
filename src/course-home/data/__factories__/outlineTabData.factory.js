import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies

import '../../../courseware/data/__factories__/courseBlocks.factory';

Factory.define('outlineTabData')
  .option('courseId', 'course-v1:edX+DemoX+Demo_Course')
  .option('host', 'http://localhost:18000')
  .attr('course_tools', ['host', 'courseId'], (host, courseId) => ({
    analytics_id: 'edx.bookmarks',
    title: 'Bookmarks',
    url: `${host}/courses/${courseId}/bookmarks/`,
  }))
  .attr('course_blocks', ['courseId'], courseId => ({
    blocks: Factory.build('courseBlocks', { courseId }).blocks,
  }))
  .attr('handouts_html', [], () => '<ul><li>Handout 1</li></ul>');
