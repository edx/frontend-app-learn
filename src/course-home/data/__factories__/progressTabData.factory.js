import { Factory } from 'rosie'; // eslint-disable-line import/no-extraneous-dependencies

// Sample data helpful when developing & testing, to see a variety of configurations.
// This set of data may not be realistic, but it is intended to demonstrate many UI results.
Factory.define('progressTabData')
  .attrs({
    certificate_data: {},
    completion_summary: {
      complete_count: 1,
      incomplete_count: 1,
      locked_count: 0,
    },
    course_grade: {
      letter_grade: null,
      percent: 0,
      is_passing: false,
    },
    section_scores: [
      {
        display_name: 'First section',
        subsections: [
          {
            assignment_type: 'Homework',
            display_name: 'First subsection',
            has_graded_assignment: true,
            num_points_earned: 0,
            num_points_possible: 1,
            percent_graded: 0.0,
            show_correctness: 'always',
            show_grades: true,
            url: 'http://learning.edx.org/course/course-v1:edX+Test+run/first_subsection',
          },
        ],
      },
      {
        display_name: 'Second section',
        subsections: [
          {
            assignment_type: 'Homework',
            display_name: 'Second subsection',
            has_graded_assignment: true,
            num_points_earned: 1,
            num_points_possible: 1,
            percent_graded: 1.0,
            show_correctness: 'always',
            show_grades: true,
            url: 'http://learning.edx.org/course/course-v1:edX+Test+run/second_subsection',
          },
        ],
      },
    ],
    enrollment_mode: 'audit',
    grading_policy: {
      assignment_policies: [
        {
          num_droppable: 1,
          short_label: 'HW',
          type: 'Homework',
          weight: 1,
        },
      ],
      grade_range: {
        pass: 0.75,
      },
    },
    studio_url: 'http://studio.edx.org/settings/grading/course-v1:edX+Test+run',
    verification_data: {
      link: null,
      status: 'none',
      status_date: null,
    },
  });
