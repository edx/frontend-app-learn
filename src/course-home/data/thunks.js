import { logError } from '@edx/frontend-platform/logging';
import { camelCaseObject } from '@edx/frontend-platform';
import {
  executePostFromPostEvent,
  getCourseHomeCourseMetadata,
  getDatesTabData,
  getOutlineTabData,
  getProgressTabData,
  postCourseDeadlines,
  postDismissWelcomeMessage,
  postRequestCert,
} from './api';

import {
  addModel,
} from '../../generic/model-store';

import {
  fetchTabFailure,
  fetchTabRequest,
  fetchTabSuccess,
  setResetDatesToast,
} from './slice';

const eventTypes = {
  POST_EVENT: 'post_event',
};

export function fetchTab(courseId, tab, getTabData) {
  return async (dispatch) => {
    dispatch(fetchTabRequest({ courseId }));
    Promise.allSettled([
      getCourseHomeCourseMetadata(courseId),
      getTabData(courseId),
    ]).then(([courseHomeCourseMetadataResult, tabDataResult]) => {
      const fetchedCourseHomeCourseMetadata = courseHomeCourseMetadataResult.status === 'fulfilled';
      const fetchedTabData = tabDataResult.status === 'fulfilled';

      if (fetchedCourseHomeCourseMetadata) {
        dispatch(addModel({
          modelType: 'courses',
          model: {
            id: courseId,
            ...courseHomeCourseMetadataResult.value,
          },
        }));
      } else {
        logError(courseHomeCourseMetadataResult.reason);
      }

      if (fetchedTabData) {
        dispatch(addModel({
          modelType: tab,
          model: {
            id: courseId,
            ...tabDataResult.value,
          },
        }));
      } else {
        logError(tabDataResult.reason);
      }

      if (fetchedCourseHomeCourseMetadata && fetchedTabData) {
        dispatch(fetchTabSuccess({ courseId }));
      } else {
        dispatch(fetchTabFailure({ courseId }));
      }
    });
  };
}

export function fetchDatesTab(courseId) {
  return fetchTab(courseId, 'dates', getDatesTabData);
}

export function fetchProgressTab(courseId) {
  return fetchTab(courseId, 'progress', getProgressTabData);
}

export function fetchOutlineTab(courseId) {
  return fetchTab(courseId, 'outline', getOutlineTabData);
}

export function resetDeadlines(courseId, getTabData) {
  return async (dispatch) => {
    postCourseDeadlines(courseId).then(response => {
      const { data } = response;
      const {
        body,
        header,
      } = data;
      dispatch(getTabData(courseId));
      dispatch(setResetDatesToast({ body, header }));
    });
  };
}

export function dismissWelcomeMessage(courseId) {
  return async () => postDismissWelcomeMessage(courseId);
}

export function requestCert(courseId) {
  return async () => postRequestCert(courseId);
}

export function processEvent(eventData, getTabData) {
  return async (dispatch) => {
    const event = camelCaseObject(eventData);
    if (event.eventName === eventTypes.POST_EVENT) {
      executePostFromPostEvent(event.postData).then(response => {
        const { data } = response;
        const {
          body,
          header,
        } = data;
        dispatch(getTabData(event.postData.bodyParams.courseId));
        dispatch(setResetDatesToast({ body, header }));
      });
    }
  };
}
