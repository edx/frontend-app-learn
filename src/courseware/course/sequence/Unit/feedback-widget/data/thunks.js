import { logError } from '@edx/frontend-platform/logging';

import {
  createWholeCourseTranslationFeedbackFailure,
  createWholeCourseTranslationFeedbackRequest,
  createWholeCourseTranslationFeedbackSuccess,
  fetchWholeCourseTranslationFeedbackFailure,
  fetchWholeCourseTranslationFeedbackRequest,
  fetchWholeCourseTranslationFeedbackSuccess,
} from './slice';
import { createTranslationFeedback, getTranslationFeedback } from './api';
import { updateModel } from '../../../../../../generic/model-store';

export function fetchWholeCourseTranslationFeedback(
  courseId,
  translationLanguage,
  unitId,
  userId,
) {
  return async (dispatch) => {
    dispatch(fetchWholeCourseTranslationFeedbackRequest({
      courseId,
    }));
    try {
      const existingFeedback = await getTranslationFeedback(
        courseId,
        translationLanguage,
        unitId,
        userId,
      );
      dispatch(updateModel({
        modelType: 'coursewareMeta',
        model: {
          id: courseId,
          translationFeedback: existingFeedback ? {
            translationLanguage,
            unitId,
            userId,
            feedbackValue: existingFeedback?.feedbackValue,
          } : null,
        },
      }));
      dispatch(fetchWholeCourseTranslationFeedbackSuccess({
        courseId,
      }));
    } catch (error) {
      logError(error);
      dispatch(fetchWholeCourseTranslationFeedbackFailure({
        courseId,
      }));
    }
  };
}

export function createWholeCourseTranslationFeedback(
  courseId,
  feedbackValue,
  translationLanguage,
  unitId,
  userId,
) {
  return async (dispatch) => {
    dispatch(createWholeCourseTranslationFeedbackRequest({
      courseId,
    }));
    try {
      const newFeedback = await createTranslationFeedback(
        courseId,
        feedbackValue,
        translationLanguage,
        unitId,
        userId,
      );
      dispatch(updateModel({
        modelType: 'coursewareMeta',
        model: {
          id: courseId,
          translationFeedback: {
            translationLanguage,
            unitId,
            userId,
            feedbackValue: newFeedback.feedbackValue,
          },
        },
      }));
      dispatch(createWholeCourseTranslationFeedbackSuccess({
        courseId,
      }));
    } catch (error) {
      logError(error);
      dispatch(createWholeCourseTranslationFeedbackFailure({
        courseId,
      }));
    }
  };
}
