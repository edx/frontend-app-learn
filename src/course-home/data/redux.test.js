import { Factory } from 'rosie';
import MockAdapter from 'axios-mock-adapter';

import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform';

import * as thunks from './thunks';

import executeThunk from '../../utils';

import './__factories__';
import '../../courseware/data/__factories__/courseMetadata.factory';
import initializeMockApp from '../../setupTest';
import initializeStore from '../../store';

const { loggingService } = initializeMockApp();

const axiosMock = new MockAdapter(getAuthenticatedHttpClient());

describe('Data layer integration tests', () => {
  const courseMetadata = Factory.build('courseMetadata');
  const courseHomeMetadata = Factory.build(
    'courseHomeMetadata', {
      course_id: courseMetadata.id,
    },
    { courseTabs: courseMetadata.tabs },
  );

  const courseId = courseMetadata.id;
  const courseBaseUrl = `${getConfig().LMS_BASE_URL}/api/courseware/course`;
  const courseMetadataBaseUrl = `${getConfig().LMS_BASE_URL}/api/course_home/v1/course_metadata`;

  const courseUrl = `${courseBaseUrl}/${courseId}`;
  const courseMetadataUrl = `${courseMetadataBaseUrl}/${courseId}`;

  let store;

  beforeEach(() => {
    axiosMock.reset();
    loggingService.logError.mockReset();

    store = initializeStore();
  });

  it('Should initialize store', () => {
    expect(store.getState()).toMatchSnapshot();
  });

  describe('Test fetchDatesTab', () => {
    const datesBaseUrl = `${getConfig().LMS_BASE_URL}/api/course_home/v1/dates`;

    it('Should fail to fetch if error occurs', async () => {
      axiosMock.onGet(courseUrl).networkError();
      axiosMock.onGet(courseMetadataUrl).networkError();
      axiosMock.onGet(`${datesBaseUrl}/${courseId}`).networkError();

      await executeThunk(thunks.fetchDatesTab(courseId), store.dispatch);

      expect(loggingService.logError).toHaveBeenCalled();
      expect(store.getState().courseHome.courseStatus).toEqual('failed');
    });

    it('Should fetch, normalize, and save metadata', async () => {
      const datesTabData = Factory.build('datesTabData');

      const datesUrl = `${datesBaseUrl}/${courseId}`;

      axiosMock.onGet(courseUrl).reply(200, courseMetadata);
      axiosMock.onGet(courseMetadataUrl).reply(200, courseHomeMetadata);
      axiosMock.onGet(datesUrl).reply(200, datesTabData);

      await executeThunk(thunks.fetchDatesTab(courseId), store.dispatch);

      const state = store.getState();
      expect(state.courseHome.courseStatus).toEqual('loaded');
      expect(state).toMatchSnapshot();
    });
  });

  describe('Test fetchOutlineTab', () => {
    const outlineBaseUrl = `${getConfig().LMS_BASE_URL}/api/course_home/v1/outline`;

    it('Should result in fetch failure if error occurs', async () => {
      axiosMock.onGet(courseUrl).networkError();
      axiosMock.onGet(courseMetadataUrl).networkError();
      axiosMock.onGet(`${outlineBaseUrl}/${courseId}`).networkError();

      await executeThunk(thunks.fetchOutlineTab(courseId), store.dispatch);

      expect(loggingService.logError).toHaveBeenCalled();
      expect(store.getState().courseHome.courseStatus).toEqual('failed');
    });

    it('Should fetch, normalize, and save metadata', async () => {
      const outlineTabData = Factory.build('outlineTabData', { courseId });

      const outlineUrl = `${outlineBaseUrl}/${courseId}`;

      axiosMock.onGet(courseUrl).reply(200, courseMetadata);
      axiosMock.onGet(courseMetadataUrl).reply(200, courseHomeMetadata);
      axiosMock.onGet(outlineUrl).reply(200, outlineTabData);

      await executeThunk(thunks.fetchOutlineTab(courseId), store.dispatch);

      const state = store.getState();
      expect(state.courseHome.courseStatus).toEqual('loaded');
      expect(state).toMatchSnapshot();
    });
  });

  describe('Test resetDeadlines', () => {
    it('Should reset course deadlines', async () => {
      const resetUrl = `${getConfig().LMS_BASE_URL}/api/course_experience/v1/reset_course_deadlines`;
      axiosMock.onPost(resetUrl).reply(201);

      const getTabDataMock = jest.fn(() => ({
        type: 'MOCK_ACTION',
      }));

      await executeThunk(thunks.resetDeadlines(courseId, getTabDataMock), store.dispatch);

      expect(axiosMock.history.post[0].url).toEqual(resetUrl);
      expect(axiosMock.history.post[0].data).toEqual(`{"course_key":"${courseId}"}`);

      expect(getTabDataMock).toHaveBeenCalledWith(courseId);
    });
  });
});
