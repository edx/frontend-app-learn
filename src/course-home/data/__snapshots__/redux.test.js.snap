// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data layer integration tests Should initialize store 1`] = `
Object {
  "courseHome": Object {
    "courseId": null,
    "courseStatus": "loading",
  },
  "courseware": Object {
    "courseId": null,
    "courseStatus": "loading",
    "sequenceId": null,
    "sequenceStatus": "loading",
  },
  "models": Object {},
}
`;

exports[`Data layer integration tests Test fetchDatesTab Should fetch, normalize, and save metadata 1`] = `
Object {
  "courseHome": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course_1",
    "courseStatus": "loaded",
  },
  "courseware": Object {
    "courseId": null,
    "courseStatus": "loading",
    "sequenceId": null,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courses": Object {
      "course-v1:edX+DemoX+Demo_Course_1": Object {
        "courseId": "course-v1:edX+DemoX+Demo_Course_1",
        "id": "course-v1:edX+DemoX+Demo_Course_1",
        "isSelfPaced": false,
        "isStaff": false,
        "number": "DemoX",
        "org": "edX",
        "tabs": Array [
          Object {
            "slug": "courseware",
            "title": "Course",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course/",
          },
          Object {
            "slug": "discussion",
            "title": "Discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/discussion/forum/",
          },
          Object {
            "slug": "wiki",
            "title": "Wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course_wiki",
          },
          Object {
            "slug": "progress",
            "title": "Progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/progress",
          },
          Object {
            "slug": "instructor",
            "title": "Instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/instructor",
          },
        ],
        "title": "Demonstration Course",
      },
    },
    "dates": Object {
      "course-v1:edX+DemoX+Demo_Course_1": Object {
        "courseDateBlocks": Array [
          Object {
            "date": "2013-02-05T05:00:00Z",
            "dateType": "course-start-date",
            "description": "",
            "extraInfo": "",
            "learnerHasAccess": true,
            "link": "",
            "title": "Course Starts",
          },
        ],
        "datesBannerInfo": Object {
          "contentTypeGatingEnabled": false,
          "missedDeadlines": false,
          "missedGatedContent": false,
        },
        "id": "course-v1:edX+DemoX+Demo_Course_1",
        "learnerIsFullAccess": true,
        "missedDeadlines": false,
        "missedGatedContent": false,
        "userTimezone": null,
        "verifiedUpgradeLink": "http://localhost:18130/basket/add/?sku=8CF08E5",
      },
    },
  },
}
`;

exports[`Data layer integration tests Test fetchOutlineTab Should fetch, normalize, and save metadata 1`] = `
Object {
  "courseHome": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course_1",
    "courseStatus": "loaded",
  },
  "courseware": Object {
    "courseId": null,
    "courseStatus": "loading",
    "sequenceId": null,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courses": Object {
      "course-v1:edX+DemoX+Demo_Course_1": Object {
        "courseId": "course-v1:edX+DemoX+Demo_Course_1",
        "id": "course-v1:edX+DemoX+Demo_Course_1",
        "isSelfPaced": false,
        "isStaff": false,
        "number": "DemoX",
        "org": "edX",
        "tabs": Array [
          Object {
            "slug": "courseware",
            "title": "Course",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course/",
          },
          Object {
            "slug": "discussion",
            "title": "Discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/discussion/forum/",
          },
          Object {
            "slug": "wiki",
            "title": "Wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course_wiki",
          },
          Object {
            "slug": "progress",
            "title": "Progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/progress",
          },
          Object {
            "slug": "instructor",
            "title": "Instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/instructor",
          },
        ],
        "title": "Demonstration Course",
      },
    },
    "outline": Object {
      "course-v1:edX+DemoX+Demo_Course_1": Object {
        "courseBlocks": Object {
          "courses": Object {
            "block-v1:edX+DemoX+Demo_Course+type@course+block@bcdabcdabcdabcdabcdabcdabcdabcd4": Object {
              "id": "course-v1:edX+DemoX+Demo_Course_1",
              "sectionIds": Array [
                "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
              ],
              "title": "bcdabcdabcdabcdabcdabcdabcdabcd4",
            },
          },
          "sections": Object {
            "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3": Object {
              "courseId": "course-v1:edX+DemoX+Demo_Course_1",
              "id": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
              "sequenceIds": Array [
                "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
              ],
              "title": "bcdabcdabcdabcdabcdabcdabcdabcd3",
            },
          },
          "sequences": Object {
            "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2": Object {
              "id": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
              "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
              "sectionId": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
              "title": "bcdabcdabcdabcdabcdabcdabcdabcd2",
              "unitIds": Array [
                "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
              ],
            },
          },
          "units": Object {
            "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1": Object {
              "graded": false,
              "id": "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
              "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
              "sequenceId": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
              "title": "bcdabcdabcdabcdabcdabcdabcdabcd1",
            },
          },
        },
        "courseTools": Object {
          "analyticsId": "edx.bookmarks",
          "title": "Bookmarks",
          "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/bookmarks/",
        },
        "datesWidget": undefined,
        "handoutsHtml": "<ul><li>Handout 1</li></ul>",
        "id": "course-v1:edX+DemoX+Demo_Course_1",
      },
    },
  },
}
`;
